sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/TypeInference.scala

```diff
@@ -0,0 +1,427 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst
+
+import java.util.{Map => JMap}
+
+import scala.collection.Map
+import scala.language.existentials
+
+import com.google.common.reflect.TypeToken
+
+import org.apache.spark.sql.catalyst.DeserializerBuildHelper._
+import org.apache.spark.sql.catalyst.JavaTypeInference._
+import org.apache.spark.sql.catalyst.ScalaReflection._
+import org.apache.spark.sql.catalyst.ScalaReflection.universe._
+import org.apache.spark.sql.catalyst.expressions.{Expression, IsNull, MapKeys, MapValues}
+import org.apache.spark.sql.catalyst.expressions.objects._
+import org.apache.spark.sql.catalyst.util.ArrayBasedMapData
+import org.apache.spark.sql.types.{ObjectType, SQLUserDefinedType, UDTRegistration, UserDefinedType}
+
+object TypeInference {
+
+  /**
+   * Returns an expression that can be used to deserialize an input expression to an object of type
+   * represented by an `TypeInference` with a compatible schema.
+   *
+   * @param t The `TypeInference` object providing type information.
+   * @param path The expression which can be used to extract serialized value.
+   * @param walkedTypePath The paths from top to bottom to access current field when deserializing.
+   */
+  def deserializerFor(
+      t: TypeInference,
+      path: Expression,
+      walkedTypePath: WalkedTypePath): Expression = t match {
+    case _ if t.isNotObjectType => path
+    case _ if t.isJavaBoolean =>
+      createDeserializerForTypesSupportValueOf(path, classOf[java.lang.Boolean])
+    case _ if t.isJavaByte =>
+      createDeserializerForTypesSupportValueOf(path, classOf[java.lang.Byte])
+    case _ if t.isJavaInteger =>
+      createDeserializerForTypesSupportValueOf(path, classOf[java.lang.Integer])
+    case _ if t.isJavaLong =>
+      createDeserializerForTypesSupportValueOf(path, classOf[java.lang.Long])
+    case _ if t.isJavaDouble =>
+      createDeserializerForTypesSupportValueOf(path, classOf[java.lang.Double])
+    case _ if t.isJavaFloat =>
+      createDeserializerForTypesSupportValueOf(path, classOf[java.lang.Float])
+    case _ if t.isJavaShort =>
+      createDeserializerForTypesSupportValueOf(path, classOf[java.lang.Short])
+    case _ if t.isJavaLocalDate =>
+      createDeserializerForLocalDate(path)
+    case _ if t.isJavaDate =>
+      createDeserializerForSqlDate(path)
+    case _ if t.isJavaInstant =>
+      createDeserializerForInstant(path)
+    case _ if t.isJavaTimestamp =>
+      createDeserializerForSqlTimestamp(path)
+    case _ if t.isJavaString =>
+      createDeserializerForString(path, returnNullable = false)
```

Comment:
At `JavaTypeInference`, it was `returnNullable = true`, but at `ScalaReflection`, it was `returnNullable = false`. This inconsistency seems weird. Let me do ``returnNullable = false` first to see if tests can be passed.

Reply:
`returnNullable` is a parameter of `Invoke`, used to indicate the invoked method will always return non-null value if it is false. I also think it is reasonable to do `returnNullable = false` here.